File: problem2-bank-account/results/JavaScript/A/attempt_1.js
'use strict';

/**
 * A simple bank account.
 *
 * - Balance starts at 0.
 * - Deposits and withdrawals must be positive integers.
 * - A withdrawal may not exceed the current balance.
 * - The balance is always >= 0.
 */
class BankAccount {
  #balance = 0;

  /**
   * Validate that an amount is a positive integer.
   * @param {number} amount
   * @param {string} op operation name, used in the error message
   */
  static #assertPositiveInteger(amount, op) {
    if (typeof amount !== 'number' || !Number.isInteger(amount)) {
      throw new Error(`${op} amount must be an integer`);
    }
    if (amount <= 0) {
      throw new Error(`${op} amount must be positive`);
    }
  }

  /**
   * Deposit a positive integer amount.
   * @param {number} amount
   */
  deposit(amount) {
    BankAccount.#assertPositiveInteger(amount, 'deposit');
    this.#balance += amount;
  }

  /**
   * Withdraw a positive integer amount that does not exceed the balance.
   * @param {number} amount
   */
  withdraw(amount) {
    BankAccount.#assertPositiveInteger(amount, 'withdraw');
    if (amount > this.#balance) {
      throw new Error('insufficient funds');
    }
    this.#balance -= amount;
  }

  /**
   * @returns {number} the current balance
   */
  balance() {
    return this.#balance;
  }
}

module.exports = { BankAccount };

File: problem2-bank-account/results/Python/A/attempt_1.py
"""A simple bank account with deposit, withdraw, and balance operations."""


class BankAccountError(Exception):
    """Raised when an invalid bank account operation is attempted."""


class BankAccount:
    """A bank account holding a non-negative integer balance."""

    def __init__(self):
        self._balance = 0

    def balance(self):
        """Return the current balance."""
        return self._balance

    def deposit(self, amount):
        """Deposit a positive integer amount into the account."""
        self._validate_amount(amount)
        self._balance += amount

    def withdraw(self, amount):
        """Withdraw a positive integer amount, not exceeding the balance."""
        self._validate_amount(amount)
        if amount > self._balance:
            raise BankAccountError("withdrawal amount exceeds balance")
        self._balance -= amount

    @staticmethod
    def _validate_amount(amount):
        if not isinstance(amount, int) or isinstance(amount, bool):
            raise BankAccountError("amount must be an integer")
        if amount <= 0:
            raise BankAccountError("amount must be a positive integer")
